QEfficient/peft/auto.py

```diff
@@ -0,0 +1,436 @@
+                json.dump(
+                    {
+                        "connections": [{"devices": list(range(mdp_ts_num_devices)), "type": "p2p"}],
+                        "partitions": [
+                            {
+                                "name": "Partition0",
+                                "devices": [{"deviceId": d, "numCores": num_cores} for d in range(mdp_ts_num_devices)],
+                            }
+                        ],
+                    },
+                    fp,
+                    indent=4,
+                )
+            command.append(f"-mdp-load-partition-config={mdp_ts_json}")
+            compile_hash.update(to_hashable({"mdp_ts_num_devices": mdp_ts_num_devices}))
+
+        # Check if already compiled
+        compile_hash = compile_hash.hexdigest()[:16]
+        qpc_path = qpc_path.with_name(qpc_path.name + "-" + compile_hash)
+        if qpc_path.is_dir():
+            if (qpc_path / "programqpc.bin").is_file():
+                self.qpc_path = qpc_path
+                return qpc_path
+            # Probably compilation failure last time, delete directory to start over
+            shutil.rmtree(qpc_path)
+
+        command.append(f"-aic-binary-dir={qpc_path}")
+        logger.info(f"Running compiler: {' '.join(command)}")
+        subprocess.run(command).check_returncode()
+
+        self.qpc_path = qpc_path
+        return qpc_path
+
+    def compile(
+        self,
+        onnx_path: Optional[str] = None,
+        compile_dir: Optional[str] = None,
+        *,
+        batch_size: int = 1,
+        prefill_seq_len: int,
+        ctx_len: int,
+        num_devices: int = 1,
+        num_cores: int = 16,
+        mxfp6_matmul: bool = False,
+        mxint8_kv_cache: bool = False,
+        **compiler_options,
+    ) -> str:
+        # Specializations
+        specializations = [
+            {"batch_size": batch_size, "seq_len": prefill_seq_len, "ctx_len": ctx_len},
+            {"batch_size": batch_size, "seq_len": 1, "ctx_len": ctx_len},
+        ]
+
+        # Custom IO
+        custom_io = {}
+        kv_cache_dtype = "mxint8" if mxint8_kv_cache else "float16"
+        for suffix in ["", "_RetainedState"]:
+            for i in range(self.num_layers):
+                for kv in ["key", "value"]:
+                    custom_io[f"past_{kv}.{i}{suffix}"] = kv_cache_dtype
+            for weight_name in self.adapter_weights[self.active_adapter]:
+                custom_io[f"{weight_name}{suffix}"] = "float16"
+
+        return self._compile(
+            onnx_path,
+            compile_dir,
+            compile_only=True,
+            retained_state=True,
+            specializations=specializations,
+            convert_to_fp16=True,
+            mxfp6_matmul=mxfp6_matmul,
+            custom_io=custom_io,
+            mdp_ts_num_devices=num_devices,
+            aic_num_cores=num_cores,
+            **compiler_options,
+        )
+
+    def generate(
```

Comment:
We provide way to generate outputs on multiple input prompts, which includes splitting inputs based on batch size as well.

The `QEffAutoModelForCausalLM` currently takes list of strings as input. We need to introduce pipelines eventually to make align our interface with transformers.

But with this PR, we have two different generate methods taking different types of inputs. So two questions.
1. How do you plan to provide way to generate inference on multiple input prompts and when?
2. Let's align this method with `QEffAutoModelForCausalLM` before the release.

If it is not possible to provide `Pipelines` before `1.18` we can make this generate align with `QEFFAutoModel` generate and provide a temporary way to generate output for multiple prompts -> which can eventually evolve into Pipelines

Reply:
Since the Peft class is a new addition, we can use a different method from AutoModelForCausalLM. Users would have to read the documentation to use the same. When we migrate AutoModelForCausalLM to the same interface, we provide that as breaking change in the release.

Since this class interface aligns more with transformers class, it's simple for users to understand.